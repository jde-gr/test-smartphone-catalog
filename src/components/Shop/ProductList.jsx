import { useState, useEffect } from 'react';
import { useQuery } from 'react-query';

import { BASE_URL } from '../../core/constants';
import ProductItem from './ProductItem';
import SearchBar from './SearchBar';
import { PLPList, PageContainer } from './styles/ProductList.styles';

const ProductList = () => {
  const fetchProducts = async () => {
    const res = await fetch(`${BASE_URL}/product`);
    const data = await res.json();
    return data;
  };

  // const queryClient = useQueryClient();
  const { status, data, error, isFetching } = useQuery(
    'products',
    fetchProducts
  );

  const [filteredData, setFilteredData] = useState([]);

  useEffect(() => {
    status === 'success' && !filteredData.length && setFilteredData(data);
  }, [status, data, filteredData]);

  const devices = filteredData?.map((device) => (
    <ProductItem
      key={device.id}
      id={device.id}
      brand={device.brand}
      model={device.model}
      price={device.price}
      imgUrl={device.imgUrl}
    />
  ));

  const filterHandler = (value) => {
    setFilteredData(
      data.filter((device) => {
        return (
          device.brand.toLowerCase().includes(value) ||
          device.model.toLowerCase().includes(value)
        );
      })
    );
  };

  return (
    <PageContainer>
      <SearchBar onChangeFilter={filterHandler} />
      {status === 'error' && <p>{error.toString()}</p>}
      {(status === 'loading' || isFetching) && <p>Fetching data...</p>}
      {status === 'success' && filteredData !== [] && (
        <PLPList>{devices}</PLPList>
      )}
    </PageContainer>
  );
};

export default ProductList;
