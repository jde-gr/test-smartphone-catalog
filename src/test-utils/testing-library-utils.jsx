import { render } from '@testing-library/react';
import { QueryClient, QueryClientProvider } from 'react-query';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import store from '../store';

const queryClient = new QueryClient();

const AllTheProviders = ({ children }) => {
  return (
    <BrowserRouter>
      <Provider store={store}>
        <QueryClientProvider client={queryClient}>
          {children}
        </QueryClientProvider>
      </Provider>
    </BrowserRouter>
  );
};
const renderWithContext = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { renderWithContext as render };
