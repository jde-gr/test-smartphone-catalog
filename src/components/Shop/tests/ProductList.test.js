import userEvent from '@testing-library/user-event';

import { render, screen } from '../../../test-utils/testing-library-utils';

import ProductList from '../ProductList';
import { productListData } from '../../../test-utils/fixtures/productListFixtures';

test('displays image for each mobile device from server', async () => {
  render(<ProductList data={productListData} />);

  // find images
  const mobileImages = await screen.findAllByRole('img', {
    name: /^dispositivo/i,
  });
  expect(mobileImages).toHaveLength(2);

  // confirm alt text of images
  const altText = mobileImages.map((element) => element.alt);
  expect(altText).toEqual([
    'Dispositivo Acer Iconia Talk S',
    'Dispositivo Acer Liquid Z6 Plus',
  ]);
});

test('displays a search bar to filter devices', () => {
  render(<ProductList data={productListData} />);

  const searchInput = screen.getByPlaceholderText(/buscar/i);
  expect(searchInput).toBeInTheDocument();
});

test('updates device list when typing on search', async () => {
  render(<ProductList data={productListData} />);

  // find device model that will be filtered and that one which will remain
  const modelText = await screen.findByText(/Iconia Talk S$/i);
  const model2Text = screen.getByText(/Liquid Z6 Plus$/i);
  expect(modelText).toBeInTheDocument();
  expect(model2Text).toBeInTheDocument();

  const searchInput = screen.getByPlaceholderText(/buscar/i);
  expect(searchInput).toBeInTheDocument();

  userEvent.clear(searchInput);
  userEvent.type(searchInput, 'liquid');

  // expect the filtered device not to appear any more, while remaining the other one
  expect(modelText).not.toBeInTheDocument();
  expect(model2Text).toBeInTheDocument();
});

test('device card links to the details page', async () => {
  render(<ProductList data={productListData} />);

  const card1Link = await screen.findByTestId('ZmGrkLRPXOTpxsU4jjAcv');
  expect(card1Link).toHaveAttribute('href', '/device/:ZmGrkLRPXOTpxsU4jjAcv');
});

/* test('handles error for the device list', async () => {
  server.resetHandlers(
    rest.get(`${BASE_URL}/product`, (req, res, ctx) => {
      return res(ctx.status(500));
    })
  );

  render(<ProductListPage />);

  await waitFor(async () => {
    const notification = await screen.findByTestId('notification');
    expect(notification).toBeInTheDocument();
  });
}); */
