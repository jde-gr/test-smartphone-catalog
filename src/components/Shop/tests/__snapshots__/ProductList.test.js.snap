// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    class="css-1sjygkg"
  >
    <div
      class="css-isrll3"
    >
      <input
        class="css-102ulqx"
        placeholder="Buscar"
        type="text"
      />
    </div>
    <div
      class="css-9yntt9"
    >
      <div
        style="width: inherit;"
      >
        <div
          class="css-1xb5uqo"
        >
          <div
            class="css-1b8ia93"
          >
            <a
              data-testid="ZmGrkLRPXOTpxsU4jjAcv"
              href="/device/:ZmGrkLRPXOTpxsU4jjAcv"
            >
              <div
                class="css-1r4ryte"
              >
                <img
                  alt="Dispositivo Acer Iconia Talk S"
                  class="css-0"
                  src="https://front-test-api.herokuapp.com/images/ZmGrkLRPXOTpxsU4jjAcv.jpg"
                />
              </div>
              <div
                class="css-ihzpm"
              >
                <h4
                  class="css-1iv4gsg"
                >
                  Marca: 
                  Acer
                </h4>
                <h6
                  class="css-bhh5v6"
                >
                  Modelo: 
                  Iconia Talk S
                </h6>
                <p
                  class="css-1kkf7xj"
                >
                  Precio: 
                  170
                  €
                </p>
              </div>
            </a>
          </div>
        </div>
      </div>
      <div
        style="width: inherit;"
      >
        <div
          class="css-1xb5uqo"
        >
          <div
            class="css-1b8ia93"
          >
            <a
              data-testid="cGjFJlmqNPIwU59AOcY8H"
              href="/device/:cGjFJlmqNPIwU59AOcY8H"
            >
              <div
                class="css-1r4ryte"
              >
                <img
                  alt="Dispositivo Acer Liquid Z6 Plus"
                  class="css-0"
                  src="https://front-test-api.herokuapp.com/images/cGjFJlmqNPIwU59AOcY8H.jpg"
                />
              </div>
              <div
                class="css-ihzpm"
              >
                <h4
                  class="css-1iv4gsg"
                >
                  Marca: 
                  Acer
                </h4>
                <h6
                  class="css-bhh5v6"
                >
                  Modelo: 
                  Liquid Z6 Plus
                </h6>
                <p
                  class="css-1kkf7xj"
                >
                  Precio: 
                  250
                  €
                </p>
              </div>
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
